### ISSUE TITLE
Add data which allows users to select types of commodities

### ISSUE THREAD

**Issue by AnaBelgun** [TerriaJS/InvestorMap#150]
For the Mines and Mineral deposits layers, users want to be able to select specific minerals to be visualised (eg. lithium, iron ore, gold etc.). 

- [ ] customer to send a list of 10-20 priority minerals

- [ ] create a new layer with minerals as displayed variables

**Comment by stevage** [comment]
Any word on the list of 10-20 priority minerals, @AnaBelgun ?

**Comment by AnaBelgun** [comment]
This is for next release; they are still deciding which minerals.

**Comment by AnaBelgun** [comment]
Re further work on the resources datasets.
As previously discussed Austrade is interested in Investor Map users also being able to load projects by specific commodities (subject to time/cost involved).
Below is an indicative list.  Some commodities could drop off if there are only a few projects.    


•	Iron ore
•	Coal
•	Gold
•	Bauxite
•	Copper 
•	Nickel
•	Lithium
•	Mineral sands (rutile,  ilmenite, zircon) 
•	Graphite
•	Cobalt
•	Manganese 
•	Uranium
•	Phosphate
•	Vanadium
•	Silver/Lead/Zinc
•	Rare Earths (REO & Y₂O₃)


Related to this would be discussing how this list correlates with the existing legend below – and how/whether this legend could be enhanced (eg whether all iron ore/iron deposits could be brought together).
![mines legend](https://user-images.githubusercontent.com/18452161/51214990-a6b00e00-1973-11e9-91a0-baf7b9014799.jpg)

**Comment by AnaBelgun** [comment]
waiting for GA to clarify with Austrade if they develop the layer (as a WMS) or we extract the data as a CSV and create a new layer with the commodities listed

**Comment by AnaBelgun** [comment]
Ideally we would replicate the Commodities with its filters: 
![image](https://user-images.githubusercontent.com/18452161/51880740-80a35880-23cc-11e9-8b7e-29e68397cf63.png)

**Comment by stevage** [comment]
Ok, so after having poked at this for a while, I don't have a great solution yet. There are several potential solutions but all with issues.

AusGIN relies on a web service, http://api.portal.geoscience.gov.au/getWMSMapViaProxy.do. For some reason, it uses a cookie,`JSESSIONID`, but maybe that doesn't matter. The bigger issue is that its URL structure is not exactly WMS:

http://api.portal.geoscience.gov.au/getWMSMapViaProxy.do?url=http%3A%2F%2Fservices.ga.gov.au%2Fearthresource%2Fows%3FSERVICE%3DWMS&version=1.3.0&layer=erl%3AMineralOccurrenceView&sldUrl=%2FmineralOccurrenceViewFilterStyle.do%3FcommodityUri%3Dhttp%253A%252F%252Fresource.geosciml.org%252Fclassifier%252Fcgi%252Fcommodity-code%252Fbrown-coal%26serviceUrl%3Dhttp%253A%252F%252Fservices.ga.gov.au%252Fearthresource%252Fows%253FSERVICE%253DWMS%2526&crs=EPSG%3A3857&bbox=15028131.257091936%2C-5009377.085697312%2C20037508.342789248%2C-9.313225746154785e-10

I'm not really sure how we'd go about supporting that. It basically has WMS-like parameters (eg, bbox, crs) but some have different names (`layer` instead of `layers`), has an implied `request=GetMap` and of course that `url` parameter.

The other big quirk is that buried in there is the logic for how it actually displays just the dots for a particular commodity: the call to `mineralOccurrenceViewFilterStyle`:

`http://api.portal.geoscience.gov.au/earthresource/mineralOccurrenceViewFilterStyle.do?commodityUri=http%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fcgi%2Fcommodity-code%2Fbrown-coal`

This is a service which returns an SLD, which is then passed by URL to GeoServer. If we don't go down the path of using the `getWMSMapViaProxy`, then we need to either:

1. Make the WMS call ourselves, passing that URL, via `sld=...`
2. Fetch the SLD ourselves, then also make the WMS call, passing the SLD in the URL, via `sld_url=...`

I tried option 1 [url](http://services.ga.gov.au/earthresource/ows?version=1.3.0&layers=erl:MineralOccurrenceView&bbox=16593561.59637234,-5009377.085697312,16906647.66422842,-4696291.01784123&SERVICE=WMS&request=GetMap&WIDTH=500&HEIGHT=500&crs=EPSG:3857&format=image/png&sld=http://api.portal.geoscience.gov.au/mineralOccurrenceViewFilterStyle.do?commodityUri=http%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fcgi%2Fcommodity-code%2Fbrown-coal&serviceUrl=http%3A%2F%2Fservices.ga.gov.au%2Fearthresource%2Fows%3FSERVICE%3DWMS%26). I got a gateway timeout (504).

I tried putting the SLD on Gist, and referring to it that way. ([url](http://services.ga.gov.au/earthresource/ows?version=1.3.0&layers=erl:MineralOccurrenceView&bbox=16593561.59637234,-5009377.085697312,16906647.66422842,-4696291.01784123&SERVICE=WMS&request=GetMap&WIDTH=500&HEIGHT=500&crs=EPSG:3857&format=image/png&sld=https://gist.githubusercontent.com/stevage/04b77cbe23db24dc18c8cf54b89e191e/raw/ca591fc736ac6cb777149cc774625c28c96a3d82/browncoal.sld)). That failed too.

So I tried option 2. With a bit of tweaking to the SLD ([url](http://services.ga.gov.au/earthresource/ows?version=1.3.0&layers=erl:CommodityResourceView&bbox=16593561.59637234,-5009377.085697312,16906647.66422842,-4696291.01784123&SERVICE=WMS&request=GetMap&WIDTH=500&HEIGHT=500&crs=EPSG:3857&format=image/png&sld_body=%3CStyledLayerDescriptor%20xmlns%3Awfs%3D%22http%3A%2F%2Fwww.opengis.net%2Fwfs%22%20xmlns%3Aogc%3D%22http%3A%2F%2Fwww.opengis.net%2Fogc%22%20xmlns%3Agml%3D%22http%3A%2F%2Fwww.opengis.net%2Fgml%22%20xmlns%3Axsi%3D%22http%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema-instance%22%20%20version%3D%221.0.0%22%20xsi%3AschemaLocation%3D%22http%3A%2F%2Fwww.opengis.net%2Fsld%20StyledLayerDescriptor.xsd%22%3E%3Cscript%2F%3E%3CNamedLayer%3E%3CName%3Eerl%3AMineralOccurrenceView%3C%2FName%3E%3CUserStyle%3E%3CName%3Eportal-style%3C%2FName%3E%3CTitle%3Eerl%3AMineralOccurrenceView%3C%2FTitle%3E%3CAbstract%3EEarthResource%3C%2FAbstract%3E%3CIsDefault%3E1%3C%2FIsDefault%3E%3CFeatureTypeStyle%3E%3CRule%3E%3CName%3Eerl%3AMineralOccurrenceView%3C%2FName%3E%3CTitle%3EMineral%20Occurrence%3C%2FTitle%3E%3CAbstract%3Eerl%3AMineralOccurrenceView%3C%2FAbstract%3E%3Cogc%3AFilter%3E%3Cogc%3APropertyIsEqualTo%20matchCase%3D%22false%22%3E%3Cogc%3APropertyName%3Eerl%3ArepresentativeCommodity_uri%3C%2Fogc%3APropertyName%3E%3Cogc%3ALiteral%3Ehttp%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fcgi%2Fcommodity-code%2Fbrown-coal%3C%2Fogc%3ALiteral%3E%3C%2Fogc%3APropertyIsEqualTo%3E%3C%2Fogc%3AFilter%3E%3CPointSymbolizer%3E%3CGraphic%3E%3CMark%3E%3CWellKnownName%3Ecircle%3C%2FWellKnownName%3E%3CFill%3E%3CCssParameter%20name%3D%22fill%22%3E%2300aa00%3C%2FCssParameter%3E%3C%2FFill%3E%3CStroke%3E%3CCssParameter%20name%3D%22stroke%22%3E%23000000%3C%2FCssParameter%3E%3CCssParameter%20name%3D%22stroke-width%22%3E0.15%3C%2FCssParameter%3E%3C%2FStroke%3E%3C%2FMark%3E%3CSize%3E8%3C%2FSize%3E%3C%2FGraphic%3E%3C%2FPointSymbolizer%3E%3C%2FRule%3E%3C%2FFeatureTypeStyle%3E%3C%2FUserStyle%3E%3C%2FNamedLayer%3E%3C%2FStyledLayerDescriptor%3E)), it eventually returned an image - although it's transparent.

I don't really know what would be required to make the images work. Part of the problem I suspect is I don't really understand the relationship between the various layers available and the attributes: how "commodity" and "resource" and "mineral occurrence" relate.

A final option would be to just give up on that and go back to making more CSVs. I'm not sure how the `mineral_deposits.csv` etc were generated, but we could make more of those, filtering appropriately.

So to summarise options:

1. Implement some custom layer type around `getWMSMapViaProxy`. @kring is there any way to do this that doesn't require adding code to TerriaJS? (Which seems bad because it's a one-off API).
2. Get the WMS and SLD combo to work, probably with help from GA. Then I think we just make a bunch of individual layers. Maybe GA could just generate and store the SLDs we need, server side, so we can just refer to them by name, and make life a lot simpler.
3. Make more CSVs. I'm not sure what the downside is here, perhaps the data source isn't as good somehow?

**Comment by AnaBelgun** [comment]
@kring to review the options 
will include in the 8th of Feb release

**Comment by meh9** [comment]
So I got it showing all the points from the WMS service: https://terria-cube.terria.io/#share=s-4qz4BfvCliAx9wTR1NuvFTSJgQE

GetFeatureInfo doesn't work though, I've not investigated that yet.

Technically I think we could already support the 2.a. option of "supporting" SLD's by creating separate layers and using the `parameters` block to specify the SLD. That's a big bunch of layers with a lot of SLD that would end up encoded in our init file though...

Option 1 doesn't seem very attractive for the reasons you gave.

Option 3 isn't great because we don't want to hold on to the data.

Note that there is also a WFS service at that URL, but it's slow and I couldn't get Terria to display anything even though it looked like it did return the data.

Out of these options I actually like 2.b. the most.

**Comment by kring** [comment]
Yeah that was my thought too @meh9: construct the SLDs at catalog build time and stuff them in the init file as parameters. It might even be possible - perhaps with a small code change - to trick terria into thinking the different SLDs are WMS dimensions, and then it would create UI on the workbench to let the user pick from them. There would be a whole lot of SLDs in the init file though. GZIP should mostly make it ok.

Not sure what you mean by 2.b though? Is that GA pre-generating the SLDs and letting us reference them by name?

**Comment by AnaBelgun** [comment]
@meh9 and @kring would you mind finalising the investigation on this? let me know if you need to contact this team at GA

**Comment by meh9** [comment]
@AnaBelgun yes, let's ask GA if they can create the SLD's as different WMS Styles on the server, which would make all WMS clients jobs easier.

**Comment by michaelsexton** [comment]
@stevage @meh9 @kring 

Michael Sexton here, technical custodian for the AusGIN Portal. I can address a few of your questions about how the portal works and what is the best way to interact with out underlying services

Firstly, the non-WMS API call made is a workaround developed to allows us to send POST WMS GetMap and GetFeatureInfo requests. The ability to POST to WMS was supported in OpenLayers 2, but removed by OpenLayers 4. In retrospect our solution is not the ideal way of doing things, it would be better to have modified our client-side interactions to allow POST requests for WMS to underlying services, but server side API calls were developed first by our partners at AuScope and we ran with that.

We use SLDs to filter the data, and there are three ways by which to do this:

1. Send the online SLD resource to the service as a URL
2. Send the SLD_BODY in a GET request
3. Send the SLD_BODY in a POST request

The first is usually not possible due to corporate sysadmins locking down port 80 on their infrastructure, the second runs into problems if the SLD_BODY gets too large. And for our purposes, the SLD_BODY can often get quite large. An example:

http://api.portal.geoscience.gov.au/mineralOccurrenceViewFilterStyle.do?commodityUri=http%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fcgi%2Fcommodity-code%2Fproduced-commodity&timescaleUri=http%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fics%2Fischart%2FPhanerozoic&serviceUrl=https%3A%2F%2Fgs.geoscience.nsw.gov.au%2Fgeoserver%2Fwfs

So the option we use to interact with our service is to POST the SLD_BODY to it. Even with a large request like the above, it is quite performant.

GetFeatureInfo for GeoServer also accepts SLD_BODY as a POST request, which means you can send the SLD generated above and interact with a point-click.

The filtering by commodity is done by sending URIs to the service from a controlled vocabulary (http://vocabs.ga.gov.au/cgi/sissvoc/commodity-code/concept.html). For your purposes you could pre-construct the SLDs you need for the commodities you want to select and POST these to our service. 

Let me know if you have any more questions.

**Comment by meh9** [comment]
@michaelsexton 

Could you explain a little more about Option 1 please?

It would be relatively trivial for us to host SLD's in an AWS S3 web bucket and send the URL to each SLD as a URL query parameter to the WMS with each access, is that how that would work, or did I misunderstand?

If that is it, have you considered (or do you already have) an online repository of pre-canned SLD's that we could just point to so we don't have to maintain our own?

**Comment by michaelsexton** [comment]
@meh9  That's how it would work, yes.

Currently our web service blocks port 80 because it is on our corporate servers. We are intending on migrating it to the cloud in the next month. This will allow sending SLDs as URLs.

It is possible for us to provide a pre-canned SLDs for you. However, if we were to go down the self-hosting route, it would be more ideal for us to set these as pre-defined styles on our service, which you simply request using the STYLES parameter.

What would be the set of commodities you would need for your requests?

**Comment by meh9** [comment]
@michaelsexton in that case I definitely think this option is the best one. :)

Of course we are much happier for you to host the SLD's - as STYLES would be preferable of course, but we can also work with URL's if you decide to host them.

Failing that we would be ok with hosting the SLD's ourselves, but then they will inevitably become outdated at some point in the future...

@AnaBelgun can you confirm a list of commodities please? Is it the list above in https://github.com/TerriaJS/InvestorMap/issues/150#issuecomment-454581545 ?

**Comment by AnaBelgun** [comment]
@michaelsexton ideally they want all commodities available in your service; they are interested in replicating as much as possible the filters for the list of commodities, categories, reserves etc.

**Comment by kring** [comment]
Bit of a long shot, but @michaelsexton is it perhaps possible to expose the three drop-downs (Commodity, Resource Category, Provider) as WMS dimensions? Not provider, I guess, because that switches between two separate WMS servers, right?

**Comment by michaelsexton** [comment]
@AnaBelgun - all the commodities available in the service is fairly extensive, there are nearly 300! (http://api.portal.geoscience.gov.au/getAllCommodities.do) 

I'm a bit unsure as to how you would implement it - our API interfaces with a vocabulary server which is the front end of a triple store that contains SKOS vocabularies. Our list of commodities are hierarchical - that is if you select 'precious metal', we use a reasoner to determine you'll want gold, silver, platinum group metals (which is also the top of a hierarchy of platinum, palladium, osmium, etc).

I had thought a set of pre-canned styles based on broad categories would solve your problem, eg 'has resources, critical commodities', 'has reserves, base metals', etc. Nothing too granular.

@kring not sure what you mean, but I think the quick answer is that it is possible with GeoServer using their `cql_filter` parameter:

https://docs.geoserver.org/stable/en/user/tutorials/cql/cql_tutorial.html

**Comment by AnaBelgun** [comment]
@michaelsexton you mean all "primary commodities", in the drop down list from agate to zirconium are 300 in total? those are the ones I was referring to. 
![image](https://user-images.githubusercontent.com/18452161/52674035-82495080-2f76-11e9-9217-5cc196f87c8e.png)

**Comment by michaelsexton** [comment]
@AnaBelgun yes, 292 to be exact, and growing.

Depending on what you need to deliver, I usually think it is preferable in most cases to group them - base metals, precious metals, rare earths, bulk commodities, energy minerals, mineral sands, etc, with some major/important commodities as standouts (copper, gold, iron ore, coal, etc).

We have the full list because we can take a broad term like "base metals" and use our reasoner to infer what the user meant:

http://api.portal.geoscience.gov.au/mineralOccurrenceViewFilterStyle.do?commodityUri=http%3A%2F%2Fresource.geosciml.org%2Fclassifier%2Fcgi%2Fcommodity-code%2Fbase-metal&serviceUrl=http://services.ga.gov.au/earthresource/wms

**Comment by AnaBelgun** [comment]
@michaelsexton ok, I'll discuss with Austrade and check this option (groups + some popular indiv. ones)

**Comment by AnaBelgun** [comment]
@michaelsexton hi. Austrade team will check the commodities list and draft a list of the most relevant/important based on your advice: have a list with the high-level groups and some popular individual ones, such as copper, gold, iron, coal etc.

To both teams (minerals and terria): @michaelsexton @meh9 @kring 
Just to confirm: Austrade would like for the users to be able to have this list of 20 max minerals and be able to select to view, single, multiple or all of them on the map. The information available for each, would include the reserves. 

They asked if possible to also show all 292 ones on a single layer as well (either part of this layer, or a separate one). They understand it won't be possible to select one, or multiple as part of the 292 layer, but part of the 10-20 Commodities layer (described above).

**Comment by AnaBelgun** [comment]
Hi @michaelsexton here is the list of the commodities from Austrade for the pre-canned styles: 
There are many ways to cut this and if GA wants/needs to regroup or combine some that is OK. If you have other suggestions in terms of grouping them, let us/Russell know please. 

•	Iron Ore 
•	Coal black
•	Coal brown
•	Gold/Silver
•	Copper
•	Nickel
•	Lead
•	Zinc
•	Bauxite/alumina
•	Lithium
•	Mineral sands (rutile,  ilmenite, zircon) 
•	Graphite
•	Cobalt
•	Manganese 
•	Uranium
•	Phosphate
•	Vanadium
•	Platinum Group Elements
•	Rare Earth Elements

**Comment by michaelsexton** [comment]
Hi @AnaBelgun 

Generating the above SLDs for commodities will be easy, where it gets hosted is another matter.

Are there any colour codes or icons you want?

I can generate the SLDs and provide them to you. It will not be able to be hosted with the service until the end of next month. If these timelines don't suit let me know.

**Comment by AnaBelgun** [comment]
HI @michaelsexton thanks. Austrade confirmed that we'll wait for the SLDs to be hosted with the service and include that in the March InvestorMap release. just to note that we'll need few days before the actual release (27-28 March) so we can build the new version and get it checked & approved.

**Comment by AnaBelgun** [comment]
from GA team: There are now styles associated with each commodity listed above. this is the service endpoint for testing
http://52.62.101.149:8080/geoserver/wms

@meh9 @kring would you like to work on this or delegate?
It needs to go in the end of March release

**Comment by stevage** [comment]
Ok, the new SLDs make life very much easier. I've set up each style as its own catalog item for consistency with other layers (rather than using the style drop down control within a single catalog item).

![Screenshot 2019-03-24 17 32 11](https://user-images.githubusercontent.com/678344/54875885-f66ef200-4e5a-11e9-9d4f-96753ed43db0.png)

That works fairly well. The only issue is that currently all the styles are circles of the same shade of magenta:

![Screenshot 2019-03-24 17 32 37](https://user-images.githubusercontent.com/678344/54875886-fa9b0f80-4e5a-11e9-875d-73ae6b03f979.png)

This will obviously be problematic for anyone who wants to layer several different kinds of resources together.

**Comment by stevage** [comment]
Also, could I have a little guidance on what the group should be called and where it should sit? Currently I made it "Mineral Resources and Reserves" (reflecting the WMS layer name that is being shown), near the top of the Mining group.

**Comment by AnaBelgun** [comment]
2019-03-25 Checking with client on: 
1) their preference of having drop down menu vs separate layers (initially was drop down selector)
2) their colour preference for those commodities (they do need to change)

**Comment by AnaBelgun** [comment]
@stevage 
re: 1) - they confirmed their preference is to have them on one layer with a drop down selector. Is that possible in the timeframe? 
2) colouring - in progress

**Comment by stevage** [comment]
Oh, sure. Heh, one layer with a drop down selector is actually much simpler than what I implemented. It also kind of removes the need for different colours because they won't be able to see several mineral types at once.

**Comment by AnaBelgun** [comment]
awesome, waiting for the final colours now and then we're ready to go
or can you implement it with what colours Michael set for now
Thank you @stevage

**Comment by AnaBelgun** [comment]
when this sub group is ready, the catalogue should change a bit for Mining:
- rename Mines and Mineral Deposits: Mines and Mineral Deposits - All and place it at the top of Mining catalogue
- name the new sub group (with this curated list): Mines and Mineral Deposits - Selected Commodities
- delete the Lithium Projects layer (included now in the new Selected group)

**Comment by stevage** [comment]
> the client wants a drop down box but also the ability to show multiple commodities at once

Hmm, I really can't imagine how that would be possible. A drop down box (or at least, this style of drop down box) is by definition "choose one out of several options". Maybe they're thinking of a drop down box that contains checkboxes - which our style selector doesn't support. (And for that matter, a single WMS layer with multiple styles simultaneously combined is not, afaik, possible.)

**Comment by AnaBelgun** [comment]
ok, we figured in the end that it would be almost impossible; please leave the implementation as is - multiple layers

**Comment by stevage** [comment]
Ok, have made all the requested changes.
